source/utf8/unchecked.h

```diff
@@ -215,51 +223,140 @@ namespace utf8
             return result;
         }
 
-        // The iterator class
+        namespace internal {
+        // The bidirectional_iterator class
         template <typename octet_iterator>
-          class iterator : public std::iterator <std::bidirectional_iterator_tag, uint32_t> { 
+          class bidirectional_iterator : public std::iterator <std::bidirectional_iterator_tag, uint32_t> { 
             octet_iterator it;
             public:
-            iterator () {}
-            explicit iterator (const octet_iterator& octet_it): it(octet_it) {}
+            bidirectional_iterator () {}
+            explicit bidirectional_iterator (const octet_iterator& octet_it): it(octet_it) {}
             // the default "big three" are OK
             octet_iterator base () const { return it; }
             uint32_t operator * () const
             {
                 octet_iterator temp = it;
                 return utf8::unchecked::next(temp);
             }
-            bool operator == (const iterator& rhs) const 
+            bool operator == (const bidirectional_iterator& rhs) const 
             { 
                 return (it == rhs.it);
             }
-            bool operator != (const iterator& rhs) const
+            bool operator != (const bidirectional_iterator& rhs) const
             {
                 return !(operator == (rhs));
             }
-            iterator& operator ++ () 
+            bidirectional_iterator& operator ++ () 
             {
                 ::std::advance(it, utf8::internal::sequence_length(it));
                 return *this;
             }
-            iterator operator ++ (int)
+            bidirectional_iterator operator ++ (int)
             {
-                iterator temp = *this;
+                bidirectional_iterator temp = *this;
                 ::std::advance(it, utf8::internal::sequence_length(it));
                 return temp;
             }  
-            iterator& operator -- ()
+            bidirectional_iterator& operator -- ()
             {
                 utf8::unchecked::prior(it);
                 return *this;
             }
-            iterator operator -- (int)
+            bidirectional_iterator operator -- (int)
             {
-                iterator temp = *this;
+                bidirectional_iterator temp = *this;
                 utf8::unchecked::prior(it);
                 return temp;
             }
-          }; // class iterator
+          }; // class bidirectional_iterator
+
+        template <typename octet_iterator>
+        class input_iterator : public std::iterator<std::input_iterator_tag, uint32_t> {
```

Comment:
Don't derive from  `std::iterator` but implement the `using`s in the class to stay future compatible.

Reply:
I derived from `std::iterator` like the previous `class iterator`.

Reply:
Sure. Just wanted to point it that `std::iterator` is now deprecated so I wouldn't add it to new code again. But well, this is minor.

Reply:
Ok `std::iterator` is deprecated in C++17. So let's get rid of it.
https://www.fluentcpp.com/2018/05/08/std-iterator-deprecated/